### ISSUE TITLE
Need to define process for making/evaluating/integrating propsals

### ISSUE THREAD

**Issue by TheSeamau5** [elm-lang/elm-plans#14]
Following comments in [this thread](https://github.com/elm-lang/elm-plans/issues/13), it has become apparent that any meaningful discussion on proposals to move Elm forward require some more rigor. 

As @evancz pointed out, let us put on hold all proposals and first find some rules that everyone can follow.

Potential Sources of Inspiration: 
[Dart language proposal template](https://github.com/dart-lang/dart_enhancement_proposals/blob/master/DEP%20Template.md)
[Dart language proposal page](https://github.com/dart-lang/dart_enhancement_proposals)

From seeing how the Dart folks do things, I infer that we would need to define: 
1. How should propsals be structured and what standards should they meet?
2. How should the review process work? Should it be a weekly or a bi-weekly or whenever process that accepts, declines, gives feedback, and decides which proposals are promising?
3. What kind of feedback should be expected and what form should it take?
4. Who should be expected to implement the proposal? At what point should a working draft of the idea appear?
5. What sorts of benchmarking/testing is needed?
6. If a proposal is accepted, when to expect the proposal to make it in a release of Elm.

Hopefully the added structure would not hurt the quality and accessibility of the proposal submission process and would instead help to make new enhancements appear faster such that the community can iterate on these ideas in a more meaningful way.

**Comment by Apanatshka** [comment]
I think it's great that you're championing this Hassan.
Here's my take on your list:
1. Is [this format](https://github.com/elm-lang/elm-compiler/blob/master/How%20to%20Request%20Features.md) good enough? Adding some kind of working prototype requirement would probably help to keep people from making too many proposals.  
2. I think bi-weekly meeting by committee that's skipped when there is no new activity is a good idea. It gives a bit more structure to these discussions. 
3. No idea, are there resources about what other languages do here? I couldn't find anything in the linked Dart stuff, although I only looked it over quickly. 
4. Preferably not Evan I'd think. Too much workload increase for Evan would become too big a factor in accept/reject. OTOH Evan has the roadmap in mind and the good ideas about when to release what feature. And I'm not sure how many people really have time to implement features they propose, or the time to keep up with the other changes if their feature is on the list of things that need to wait for a later version of the language.  
   I think a working draft should really be the responsibility of the proposal author or another supported of the proposal who has more time. 
5. That depends on the proposal. It's also a problem that many things don't have existing tests/benchmarks yet. 
6. That would be Evan's call.

**Comment by evancz** [comment]
I think one thing to think about is that 99% of the time, proposals should be rejected. This is how Python stayed excellent for 20 years. Saying "you can implement it" is weird because I don't think that actually changes the viability of the idea.

Maybe the process can be geared more towards "how do we turn this proposal into something that can be accomplished without changing the core language?" Maybe that means having something like OCamlp4 or the elmsprima idea that @TheSeamau5 brought up a while ago.

I think prototyping could be valuable in certain scenarios though. If it's something that needs benchmarks, those can probably happen in advance of needing my direct focus. Or if it is a risky idea that would be cool if it works, but maybe won't work.

I think it'd also be good to do a summarization process periodically. Maybe we can have an issue open cataloging all the different ideas on X, instead of 10 issues open about various aspects of it.

**Comment by mgold** [comment]
99%? I'd expect closer to 80-90%...

I was thinking of doing a summarization process on all the "rough edges" with publishing a package. The problem is that this is really better suited to tags or milestones, not a list of issues in a separate repo.

**Comment by rtfeldman** [comment]
I have a different perspective on changes to the core language, which I'd like to at least articulate.

I see Elm as a language that is aiming high in treacherous ground. As [Evan's CurryOn keynote](https://www.youtube.com/watch?v=oYk8CKH7OhE) pointed out, purely functional languages have been around for decades, but exactly zero have successfully become mainstream...in all of history!

In 2009, Haskell - the most popular purely functional programming language in history - had been around for almost two decades. Suffice it to say, it was stable. Yet instead of picking it up and using it, people wrote Go from scratch.

Like, Go. Haskell was old enough to vote, and people wrote Go from scratch instead. [Go](https://www.google.com/search?site=&tbm=isch&source=hp&biw=1440&bih=801&q=go+gopher&oq=go+gopher&gs_l=img.3..0l2j0i30j0i5i30l3j0i8i30l2j0i24l2.276.1550.0.1740.9.8.0.0.0.0.130.603.5j2.7.0....0...1ac.1.64.img..2.7.601.KqXwTYICiJ8).

And how did that turn out? Today [Go has more GitHub activity than Haskell](http://adambard.com/blog/top-github-languages-2014/).

In all fairness, though...if you're a C++ programmer, which of these homepage code snippets do you gravitate towards?

[Option 1](https://www.haskell.org/)

``` haskell
primes = filterPrime [2..] 
  where filterPrime (p:xs) = 
          p : filterPrime [x | x <- xs, x `mod` p /= 0]
```

[Option 2](https://golang.org/)

``` go
// You can edit this code!
// Click here and start typing.
package main

import "fmt"

func main() {
    fmt.Println("Hello, 世界")
}
```

(That's not entirely fair, since in 2009 Haskell's homepage looked [like this](https://ladygeekgirl.files.wordpress.com/2014/11/raiders-of-the-lost-ark-nazis.jpg). But still.)

What this teaches me is that industry programmers don't just value familiarity, they **love** familiarity. A scratch-built incremental improvement over the status quo (Go over C++) can mop the floor with a well-established paradigm-shifting improvement, if the incumbent seems hard to learn and the challenger seems familiar.

I think Elm is a huge improvement over JavaScript, Go, and for that matter Haskell. I can't get enough of it. But I'm easy! I love using top-shelf tools, and am willing to stray far from the beaten path in pursuit of them. All Elm has to do to get me to use it is to be a great language once you get comfortable with it, and it not only meets that bar but absolutely crushes it. However, for most programmers that isn't enough.

Elm is a great language, but for most industry programmers, it's not a familiar language. Since it's a great language in large part _because_ of its differences from most mainstream languages (mutable state would be more familiar, for example), a lot of the unfamiliarity is unavoidable.

To my mind, then, there are two ways of increasing the likelihood that Elm becomes the first purely functional programming language to go mainstream:
1. Become more familiar in ways that don't sacrifice anything important. (e.g. offering string interpolation, styling commas at the end of lines, etc.)
2. Offer UX improvements that are so desirable they motivate people to overcome their natural inclination to avoid the unfamiliar. (e.g. the best package manager in the history of programming, a time-traveling debugger, etc.)

I agree that once a language has done a 1.0 release, stability should become a serious priority. Let's definitely be like Python at that point!

However, I think pre-1.0, at least exploring ideas is very important. An incremental improvement over C++ can take off, but I don't think an incremental improvement over Haskell can. And while Elm is pretty far from an incremental improvement to Haskell in my book, I think too many industry programmers still look at both Elm and Haskell side-by-side and say "yeah these are just too far out there."

I'd hate for us to miss out on potential major improvements to that situation because we want to avoid changing the core language. Right now we have low (but growing!) adoption, which means breaking changes aren't a huge deal. It's true that all the other costs apply just as much as ever - @evancz's time comes first to mind - but the cost of breaking changes is a cost that only goes up with time, and I think we should take advantage of its current low point to explore before it grows prohibitively high.

My two cents. :smiley:

**Comment by TheSeamau5** [comment]
@rtfeldman I agree with a lot of your points, but I'd like to submit a counterexample to the familiarity argument: Clojure. 

How did that get so much traction? It's a Lisp, which means parentheses, parentheses, parentheses... Don't get me wrong, I think that Clojure is an amazing language and is right up there with Haskell and so on... But, it's far from familiar: It's functional, it has parentheses everywhere, it emphasizes immutability, it has crazy macros, it has parentheses everywhere, it's a legit typed functional language if you use core.typed, and did I mention it has parentheses everywhere?

I look at Clojure and I see that being unfamiliar didn't hurt it in anyway. So, how come it has more adoption than Haskell even when being much younger and thus less stable? My opinion, it succeeded because: 
1. Unlike Haskell, Clojure mainly runs on already existing mainstream platforms (JVM, CLR, JS)
2. Unlike Haskell, Clojure doesn't speak of categories. Obviously, like most languages, the core libraries are filled with Functors and Monads and all those wonderful things. Yet, Clojure just says that it's functional and immutable. No scary terminology. (except for Transducers...)
3. Rich Hickey and David Nolen are good at inspiring people during talks.
4. And perhaps more importantly, Clojure was made first and foremost to solve actual problems and made to be actually used. Haskell was originally (and still is to a certain extent) a lab for experiments. That's why in Haskell there's a bajillion language extensions. 

Looking at this, if Elm were to take cues from Clojure, it would be to make improvements only in the service of whatever goal Elm wishes to accomplish (which I understand to be making front end programming nicer and more reliable and scalable). For example, string interpolation is heavily used in front end programming... the DOM is full of strings. Were Elm targeting iOS, this may not be necessary as Cocoa is not filled with strings. But Elm targets Html5, and thus the DOM. So any improvements to make the experience of interacting with the DOM more pleasant, should be prioritized, in my opinion.

That said, @evancz is the one responsible for the language and growing it. I'm sure he sees tons of great language features that we've not even heard of that he'd like to add to Elm but refrains from doing so.

**Comment by rtfeldman** [comment]
Great analysis! I totally agree, and I definitely think Clojure is an inspiring success story for FP in the 21st century.

I'd say Clojure faced similar familiarity problems from being a Lisp as what we do from being statically typed and having no side effects. I don't know which set of challenges is harder to overcome, but certainly we didn't see any Lisps taking off before Clojure, just like today we have yet to see any purely functional languages take off.

I also think Clojure is a good example of being constrained by its selling points (Lisp macros, immutability, etc) bringing unavoidable alienness to the language, and overcoming that with a nice UX.

People comparing Clojure to Java, which it relates to kinda like how Elm relates to JavaScript, can identify a lot of UX improvements compared to Java: simplicity, conciseness, first class functions (which Java lacked when Clojure came out), a REPL...

That said, I think my original point still stands: we still have a big uphill battle, and I think now is the right time to discuss big breaking changes - regardless of whether they make things more or less familiar! - because once things progress past a certain point, the cost of those changes goes way up.

I will be very happy if Elm's first few years of industry adoption go like Clojure's. :D

**Comment by mgold** [comment]
I think Clojure is still seen as somewhat niche. It's not an academic niche, like Haskell, but it's a high-concurrency big-data performance-matters niche. You're not going to start a startup with Clojure, despite what _Beating the Averages_ tells you. You use Clojure because you want to do something different, or because you have to.

It's only anecdotal, but I worked for a company that used to be a Rails and JS shop and moved (fairly abruptly) to Clojure. Let's just say that it did not go well.

But that aside, I do agree with Richard in the importance of refining the language. If you think the cost of change is high now, it will only get higher. (The cost of change, which is retraining people and rewriting code, is different from the cost of making the change, the implementation, though that may get more difficult too as the language expands.)

It's easy to get so good at using a technology that you become blind to its weak points. Here, I think we need to borrow some UX principles: 5 new users will find 80-90% of your UX problems. If they all report the same problem, you need to fix it, but if they all report different problems, you can safely ignore all of them. When people file issues on Github that are dupes, the response is typically "we'll get around to it eventually". But someone took the time to file the issue, and they encountered the issue in their code, so it means that it's a UX problem. We have a [huge thread](https://groups.google.com/forum/#!topic/elm-discuss/J5dBfWJoLoQ) of people confused about mailboxes, and that indicates we need to change the names/metaphor, change the types, write some documentation, or some combination of those.

We're all here to help, volunteering our time, and it sounds a little disheartening to say that 99% of proposals are _a priori_ not worth implementing.

**Comment by evancz** [comment]
I don't know the actual percentage of things that go in, but that's not the point I was trying to make.

I'm saying that if I took every suggestion I got, this language would be crap. Every day people tell me what is _really_ important to work on. Sometimes it is public emails, sometimes it is chats at all hours of the day, sometimes it is in person, sometimes it is on twitter, sometimes it is a house guest, sometimes on github, sometimes in private email, sometimes in text from friends, sometimes my mom points out things that can be better, sometimes it is folks telling me I should take more time off. Every day people tell me how they would do it, what needs attention right now, what could be better. That's all fine, that's part of the process. It's the data, and it's my job to handle this in a good way and balance it with my life.

Point is, focusing on one thing means most people are waiting for something else. I think @rtfeldman and @mgold are both upset at me to some degree right now because they are waiting on a thing they care about. Janis is some sort of mad at me because I'm taking forever on simple package.elm-lang.org issues. I'm not saying these things are not important or that they won't get in. I'm saying I can't do everything at the same time, and if I tried, it would be terrible for Elm. I'm also saying the process we have to deal with these things is not working. We need to turn talk into distributed work.

This is probably the most frustrating part. Most of the time, I totally understand people's concerns. Often I have thought it through before it comes up and have a plan. I really get the address thing. The frustrating thing is that I got it a month ago. I got it months before that too when it was called `Channel` and when it was called `Input`. I get the thing about recursive data structures and working with JSON. I'd love to solve these things right this second. It feels like crap to have folks who are passionate about this project have important concerns _and_ want to address those concerns _and_ be unable to do it right now. My current daily hours are "waking" and I don't seem to take days off these days.

The point I was actually trying to make above though was that having a prototype implementation does not directly translate into it getting added into the language. I don't want to set up bad expectations for folks where they think "if I drop a big PR and follow the rules, it has to get in." I think that's a totally reasonable thing to say: making a PR does not mean it happens. That's what I was trying to say above.

This issues is about designing a process so threads don't go crazy. Even in such a thread we get very very side tracked. I think the issue with this repo is that there is no true owner who knows "this is what we do here". When it was suggested to me, two people volunteered to be admins. I think I was explicit in saying I can't manage something like this effectively in addition to my other work, but maybe I misremember. If I didn't say it then, now we know.

In any case, so far it's the same discussions we normally have had on the mailing list. That's fine, but that is not the original intent of this thing as I understood it. If the goal is, "how do we organize folks into working on useful core projects?" I don't really see this working right now.

Please lets put this on hold until @rtfeldman talk about this on Wednesday.

**Comment by rtfeldman** [comment]
Argh, I need a better personal filter with stuff like this. My sincere, heartfelt apologies; if I gave you the impression I was upset about prioritization, or progress on a particular aspect of the language, I guarantee it was some combination of wanting to make sure feedback didn't get lost while it was on my mind...and, if I'm being honest with myself, garden variety venting.

That's on me. Really sorry about that. I gotta be more mindful about piling on. I know you get it.

On the record, I think Elm is in the best possible hands it could be. I've said this to several people in private, but it bears saying publicly too. You do a crazy excellent job at both shipping improvements to the language and - especially - not pulling the trigger prematurely on changes. My trust in your judgment on these things would be hard-pressed to increase.

Anyway, tl;dr sorry for miscommunicating and thanks for everything you do to make Elm the best language I've ever used. Seriously. :)

**Comment by mgold** [comment]
I'll echo Richard's sentiment. Sorry for being overzealous, and thank you for acknowledging that we might be "upset" at you -- I'd say mildly frustrated, and much more mildly after reading that post!

**Comment by evancz** [comment]
In the process of sorting this out. Part of this process is closing down this repo.

I don't think it works to discuss design decisions in a totally unguided format like this. Ultimately, I am going to chose things and do not want to be adding a significant amount of additional stuff to the language. The stuff I do want to add, I will manage the process for that on my own, talking with contributors and such in a more structured way.
